graph/appender/dead_service.go

```diff
@@ -0,0 +1,49 @@
+package appender
+
+import (
+	"strings"
+
+	"github.com/kiali/kiali/graph/tree"
+	"github.com/kiali/kiali/kubernetes"
+)
+
+// DeadServiceAppender is responsible for removing from the graph any service nodes for which
+// the service is undefined (presumabley has been removed from K8S) and for which there is
+// no traffic reported. (kiali-621)
+type DeadServiceAppender struct{}
+
+func (a DeadServiceAppender) AppendGraph(trees *[]tree.ServiceNode, namespaceName string) {
+	istioClient, err := kubernetes.NewClient()
+	checkError(err)
+
+	for _, tree := range *trees {
+		applyDeadServices(&tree, namespaceName, istioClient)
+	}
+}
+
+func applyDeadServices(n *tree.ServiceNode, namespaceName string, istioClient *kubernetes.IstioClient) {
+	// set children to list filtered of dead services
+	filteredChildren := make([]*tree.ServiceNode, 0)
+	for _, child := range n.Children {
+		isDead := false
+		rate, hasRate := child.Metadata["rate"]
+		if hasRate && rate.(float64) == 0 {
+			// determine if there is a backing service for this node
+			serviceDetails, err := istioClient.GetServiceDetails(namespaceName, strings.Split(n.Name, ".")[0])
```

Comment:
Jumping here because I'm coding KIALI-636 and I guess we need to unify appenders or change the appenders architecture.

I mean, each separate appender makes a call into the backend, and then try to identify the condition to modify the graph, iterating per service.

But, when we have separated appenders, then each one makes a call, and I think that it is duplicating, just because, appenders are independent.

So, without enter in re-arch, I guess it makes sense to use a single appender, fetch the details for service, and check, circuit breakers, route rules, and other things, as per service we only need to fetch info once.

Reply:
@lucasponce If the single call can provide the information needed then this filtering could feasibly be combined with other badging, etc.  Although, I think this could be useful as the first appender, to filter away dead services before subsequent appenders get called.  It should not be too heavy because it will inquire only about services with no traffic, and it would not use GetServiceDetails but instead use directly the call like getServicePods.